src/action/list_directory.rs

```diff
@@ -0,0 +1,727 @@
+        use ParseError::*;
+
+        match *self {
+            UnsupportedValue(ref field, ref value) => {
+                write!(fmt, "value {} in {} field is not supported",
+                       value, field)
+            }
+        }
+    }
+}
+
+impl From<ParseError> for session::Error {
+
+    fn from(error: ParseError) -> session::Error {
+        session::Error::action(error)
+    }
+}
+
+
+/// A response type for the list directory action.
+pub struct Response {
+    mode: Option<u64>,
+    ino: Option<u32>,
+    dev: Option<u32>,
+    nlink: Option<u32>,
+    uid: Option<u32>,
+    gid: Option<u32>,
+    size: Option<u64>,
+    atime: Option<SystemTime>,
+    mtime: Option<SystemTime>,
+    ctime: Option<SystemTime>,
+    blocks: Option<u32>,
+    blksize: Option<u32>,
+    rdev: Option<u32>,
+    flags_linux: Option<u32>,
+    symlink: Option<String>,
+    pathspec: PathSpec,
+    crtime: Option<SystemTime>,
+}
+
+impl Default for Response {
+
+    fn default() -> Response {
+        Response {
+            mode: None,
+            ino: None,
+            dev: None,
+            nlink: None,
+            uid: None,
+            gid: None,
+            size: None,
+            atime: None,
+            mtime: None,
+            ctime: None,
+            blocks: None,
+            blksize: None,
+            rdev: None,
+            flags_linux: None,
+            symlink: None,
+            pathspec: PathSpec {
+                path_options: None,
+                path: Default::default(),
+            },
+            crtime: None,
+        }
+    }
+}
+
+/// A request type for the list directory action.
+pub struct Request {
+    pathspec: PathSpec,
+}
+
+enum PathOption {
+    CaseInsensitive,
+    CaseLiteral,
+    Regex,
+    Recursive,
+}
+
+struct PathSpec {
+    path_options: Option<PathOption>,
+    path: PathBuf,
+}
+
+fn get_accesses_time(metadata: &Metadata) -> Option<SystemTime> {
+    match metadata.accessed() {
+        Ok(atime) => Some(atime),
+        Err(err) => {
+            warn!("unable to get last access time: {}", err);
+            None
+        }
+    }
+}
+
+fn get_modification_time(metadata: &Metadata) -> Option<SystemTime> {
+    match metadata.modified() {
+        Ok(atime) => Some(atime),
+        Err(err) => {
+            warn!("unable to get last modification time: {}", err);
+            None
+        }
+    }
+}
+
+fn get_creation_time(metadata: &Metadata) -> Option<SystemTime> {
+    match metadata.created() {
+        Ok(atime) => Some(atime),
+        Err(err) => {
+            warn!("unable to get creation time: {}", err);
+            None
+        }
+    }
+}
+
+
+#[cfg(target_os = "linux")]
+fn get_status_change_time(metadata: &Metadata) -> Option<SystemTime> {
+    UNIX_EPOCH.checked_add(Duration::from_secs(metadata.ctime() as u64))
+}
+
+
+#[cfg(target_os = "linux")]
+fn fill_response(metadata: &Metadata, file_path: &PathBuf) -> Response {
```

Comment:
Borrowing a `PathBuf` is not very idiomatic (the same as borrowing `String`). Prefer to use `Path` or `AsPath` trait instead.